src/fdc/DiskImageUtils.cc

```diff
@@ -500,21 +526,42 @@ static void partitionNextor(SectorAccessibleDisk& disk, std::span<const unsigned
 
 		ptSector += size;
 	}
+	return narrow<unsigned>(clampedSizes.size());
 }
 
-static void partitionSunrise(SectorAccessibleDisk& disk, std::span<const unsigned> sizes)
+static unsigned partitionSunrise(SectorAccessibleDisk& disk, std::span<const unsigned> sizes)
 {
 	SectorBuffer buf;
 	ranges::fill(buf.raw, 0);
 	auto& pt = buf.ptSunrise;
 
-	assert(sizes.size() <= pt.part.size());
+	std::vector<unsigned> clampedSizes;
+	{
+		size_t sizeRemaining = std::min(disk.getNbSectors(), size_t(std::numeric_limits<unsigned>::max()));
+		if (sizeRemaining == 0) {
+			return 0;
+		}
+		--sizeRemaining;
+		for (auto [i, size] : enumerate(sizes)) {
+			if (i >= pt.part.size()) {
+				break;
+			} else if (sizeRemaining > size) {
+				clampedSizes.push_back(size);
+				sizeRemaining -= size;
+			} else if (sizeRemaining > 0) {
+				clampedSizes.push_back(narrow<unsigned>(sizeRemaining));
+				break;
+			} else {
+				break;
+			}
+		}
+	}
```

Comment:
Could the calculation of `clampedSizes` be generalized and reused between `partitionNextor()` and `partitionSunrise()`?